src/Microsoft.Diagnostics.Monitoring.EventPipe/Counters/EventCounterPipeline.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Diagnostics.NETCore.Client;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Diagnostics.Monitoring.EventPipe
+{
+    public class EventCounterPipeline : EventSourcePipeline<EventPipeCounterPipelineSettings>
+    {
+        private readonly IEnumerable<IMetricsLogger> _metricsLogger;
+        private readonly CounterFilter _filter;
+
+        public EventCounterPipeline(DiagnosticsClient client,
+            EventPipeCounterPipelineSettings settings,
+            IEnumerable<IMetricsLogger> metricsLogger) : base(client, settings)
+        {
+            _metricsLogger = metricsLogger ?? throw new ArgumentNullException(nameof(metricsLogger));
+
+            if (settings.CounterGroups.Length > 0)
+            {
+                _filter = new CounterFilter();
+                foreach (var counterGroup in settings.CounterGroups)
+                {
+                    _filter.AddFilter(counterGroup.ProviderName, counterGroup.CounterNames);
+                }
+            }
+            else
+            {
+                _filter = CounterFilter.AllCounters;
```

Comment:
I don't think the CounterFilter.AllCounters works here as I expect. It has 0 providers, so the pipeline doesn't actually request anything from the target process other than probably rundown information.

Reply:
The current behavior is for the configuration to request data from the filters. If no filtering data is provided, it requests the default set of providers. If filtering data exists, it will request only those providers.

Reply:
`MetricsService.ExecuteAsync` provides empty array of `EventPipeCounterGroup` on `CounterGroups` property. Thus, `EventCounterPipeline.ctor` sets `_filter` to `CounterFilter.AllCounters`. In `EventCounterPipeline.CreateConfiguration`, it passes `_filter` into a new `DiagnosticsEventPipeProcess`, which just calls the `CounterFilter.GetProviders` method to get the list of providers. The `CounterFilter.GetProviders` method returns `_enabledCounters.Keys`. The `CounterFilter.AllCounters` static property is just an instance of `CounterFilter` with no additional configuration, thus its `_enabledCounters` field is an empty dictionary. That means `CounterFilter.AllCounters.GetProviders()` returns an empty enumerable. So the metrics service starts the pipeline with no providers.

Reply:
The default provider creation is inside `MetricSourceConfiguration`. I can push this earlier, for example to the counter filter itself so that the filter acts more like a specification.